### DISCUSSION TITLE
Transparent line drawing issue

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#5269] [Help] [answered]
**Why is there an opaque 1 px line in the center of my half transparent line on IM 7.1.0.39 Q16 Mac OSX Monterey?**

`magick -size 500x500 xc:white -fill "rgba(255,0,0,0.5)" -strokewidth 10 -stroke "rgba(255,0,0,0.5)" -draw "line 0,0 499,499" x.png`

![x](https://user-images.githubusercontent.com/13458218/175570741-7d956520-b6cf-4b19-9a3a-b339000f09c5.png)

If I remove the fill color, it is a black line

`magick -size 500x500 xc:white -strokewidth 10 -stroke "rgba(255,0,0,0.5)" -draw "line 0,0 499,499" y.png`

![y](https://user-images.githubusercontent.com/13458218/175570933-e3ea4255-7430-418e-bc0b-3e90ff0c1ea9.png)

Would someone try to confirm this issue?

Looks to me like an issue with stokewidth

**Comment by snibgo** [comment]
I don't understand what the issue is.

All the output pixels are opaque. This is because the canvas ("xc:") is opaque, and drawing any colours (including semi-transparent colours) on opaque pixels makes opaque results.

The stroke fully overlaps the fill, so that complicates the situation.

**Reply by fmw42** [reply]
The color is reduced by the alpha value = 0.5 as expected. But why is there a full red value single px line down the center line? My fill color is the same as the stroke color, but they show up differently. Also if I remove the fill, I get a black line down the center. So it looks like the stroke is not filling the full width of the line.

**Comment by snibgo** [comment]
The fill is only one pixel wide (plus some anti-aliasing). If you don't specify the fill colour, it is opaque black. If you don't want a fill, use "-fill none".

The stroke is 10 pixels wide. If you don't specify a stroke colour, this is equivalent to "-stroke none".

When you have both fill and stroke, the pixels in the centre are drawn twice, with both the fill and the stroke.

**Reply by fmw42** [reply]
But why should it not all be the same color even if drawn twice. I would not have expected to see a different color for the center line than the rest of the line.  Looks like to achieve what I had expected, I need to set the fill color to none.

`magick -size 500x500 xc:white -fill "none" -strokewidth 10 -stroke "rgba(255,0,0,0.5)" -draw "line 0,0 499,499" x2.png`

![x2](https://user-images.githubusercontent.com/13458218/175612008-1b3ec7e4-afd5-48a7-b8d9-fd57f5203275.png)

**Comment by snibgo** [comment, marked as answer]
The fill and stroke are the same colour (same RGB values), but both are semi-transparent. Drawing one semi-transparent red over another semi-transparent red strengthens the colour.

For example (Windows syntax):
```
magick ^
  -size 300x300 xc:White ^
  -fill "rgba(255,0,0,0.5)" ^
  -draw "rectangle 10,150,290,170" ^
  -draw "rectangle 150,10,170,290" ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/175613816-8b1db3b6-02da-4683-8570-9c45fd9a293c.png)

**Reply by fmw42** [reply]
Thanks. I would have thought that two semi-transparent overlapping colors would have had an effectively smaller (or half) combined alpha value (0.5 * 0.5 = 0.25). I do not really understand alpha blending.

**Comment by snibgo** [comment]
Alpha is opacity, the proportion of light that is *not* transmitted. Suppose we have two objects, each with opacity 0.5. Put one in front of the other. The combination can transmit less light than either one alone. So the opacity (the alpha) increases.

If we repeat the previous test, with a "None" canvas:
```
%IMG7%magick ^
  -size 300x300 xc:None ^
  -fill "rgba(255,0,0,0.5)" ^
  -draw "rectangle 10,150,290,170" ^
  -draw "rectangle 150,10,170,290" ^
  x2.png
```
At the intersection, we have *transparency* of 0.5 of 0.5, which is 0.25. So the alpha (opacity) is 0.75.
![x2](https://user-images.githubusercontent.com/33812368/175646715-8ba16e18-3d2c-4494-9d3d-f063e45b6131.png)

**Reply by fmw42** [reply]
Got it, thanks.  I was thinking alpha and should have been thinking opacity.  Great explanation.
